File: 2025-1/sesion-de-clase-09/SoftProg/SoftProgDBManager/src/main/java/pe/edu/pucp/inf30/softprog/config/DBManager.java
package pe.edu.pucp.inf30.softprog.config;

import java.io.IOException;
import java.io.InputStream;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Properties;

public class DBManager {
    private static DBManager dbManager;
    
    private String host;
    private int puerto;
    private String esquema;
    private String usuario;
    private String password;
    
    private DBManager() throws IOException {
        cargarProperties();
    }
    
    public synchronized static DBManager getInstance() throws IOException {
        if (dbManager == null) {
            createInstance();
        }
        return dbManager;
    }
    
    private static void createInstance() throws IOException {
        dbManager = new DBManager();
    }
    
    public Connection getConnection() throws SQLException, ClassNotFoundException {
        try {
            /* 
            Por ahora creamos una conexion cada vez que se necesita acceder a la base de datos, 
            por ser una aplicacion academica es una practica aceptable, en un sistema productivo
            se debe usar un pool de conexiones.
            */
            Class.forName("com.mysql.cj.jdbc.Driver");
            String cadenaConexion = cadenaConexion(host, puerto, esquema);
            return DriverManager.getConnection(cadenaConexion, usuario, password);
        }
        catch (ClassNotFoundException | SQLException e) {
            System.err.println(e);
            throw e;
        }
    }
    
    private void cargarProperties() throws IOException {
        Properties properties = new Properties();
        try (InputStream input = getClass().getClassLoader().getResourceAsStream("db.properties")) {
            if (input == null) {
                System.err.println("No se pudo abrir el archivo db.properties");
                return;
            }
            
            properties.load(input);
            
            host = properties.getProperty("db.host");
            puerto = Integer.parseInt(properties.getProperty("db.puerto"));
            esquema = properties.getProperty("db.esquema");
            usuario = properties.getProperty("db.usuario");
            password = properties.getProperty("db.password");
        }
        catch (IOException e) {
            System.err.println("No se pudo cargar el archivo db.properties");
            throw e;
        }
    }
    
    private String cadenaConexion(String host, int puerto, String esquema) {
        return String.format("jdbc:mysql://%s:%d/%s?useSSL=false&allowPublicKeyRetrieval=true", host, puerto, esquema);
    }
}

File: 2025-1/sesion-de-clase-10/SoftProgDBManager/DBManager.cs
using System;
using System.Data.SqlClient;
using System.Configuration;

namespace PUCP.Edu.Pe.SoftProg.Config {
    public class DBManager {
        private static DBManager instance;

        private string host;
        private int puerto;
        private string baseDatos;
        private string usuario;
        private string password;

        private DBManager() { 
        }

        public static DBManager GetInstance() {
            if (instance == null) {
                instance = new DBManager();
            }
            return instance;
        }

        public SqlConnection GetConnection() {
            try {
                string cadenaConexion = this.CadenaConexion();
                var conexion = new SqlConnection(cadenaConexion);
                conexion.Open(); // Necesario en C#
                return conexion;
            }
            catch (SqlException e) {
                Console.Error.WriteLine(e);
                throw;
            }
        }

        private string CadenaConexion() {
            this.host = ConfigurationManager.AppSettings["db.host"];
            this.puerto = int.Parse(ConfigurationManager.AppSettings["db.puerto"]);
            this.baseDatos = ConfigurationManager.AppSettings["db.basedatos"];
            this.usuario = ConfigurationManager.AppSettings["db.usuario"];
            this.password = ConfigurationManager.AppSettings["db.password"];

            SqlConnectionStringBuilder csBuilder = new SqlConnectionStringBuilder() { 
                DataSource = $"{this.host},{this.puerto}",
                InitialCatalog = this.baseDatos,
                UserID = this.usuario,
                Password = this.password,
                TrustServerCertificate = true
            };

            return csBuilder.ConnectionString;
        }
    }
}
